source/base/optional.h

```diff
@@ -0,0 +1,131 @@
+/**
+ *  @note This file is part of Empirical, https://github.com/devosoft/Empirical
+ *  @copyright Copyright (C) Michigan State University, MIT Software license; see doc/LICENSE.md
+ *  @date 2020.
+ *
+ *  @file optional.h
+ *  @brief Audited implementation of std::optional.
+ *  @note Status: RELEASE
+ *
+ *  In release mode, functions identically to std::optional.
+ *  In debug mode, operator * and operator-> value accesses are
+ *  checked for undefined behavior.
+ */
+
+#ifndef EMP_OPTIONAL_H
+#define EMP_OPTIONAL_H
+
+#include <type_traits>
+#include <utility>
+
+#include "assert.h"
+
+#ifdef EMP_NDEBUG
+
+namespace emp {
+  template <typename T> using optional = std::optional<T>;
+
+  template <class T>
+  inline decltype(auto) make_optional(T&& val) {
+    return std::make_optional<T>(std::forward<T>(val));
+  }
+
+  template <class T, class... Args>
+  inline decltype(auto) make_optional(Args&&... args) {
+    return std::make_optional<T>(std::forward<Args>(args)... );
+  }
+
+}
+
+#else // #EMP_NDEBUG *not* set
+
+namespace emp {
+
+  template<class T>
+  class optional : public std::optional<T> {
```

Comment:
this looks great!
since we're going the emp_assert route, returning operator->() instead of value() (which _would_ throw a std::bad_optional_access) here makes sense

Reply:
this is actually a great idea I hadn't considered. I agree --- let's just do value()